### ISSUE TITLE
Include Activity Outcome & Output in execution log

### ISSUE THREAD

**Issue by sfmskywalker** [elsa-workflows/elsa-core#594]

**Comment by froens** [comment]
Great! This would be really nice! :)

**Comment by mohdali** [comment]
Execution log needs to be more informative indeed. And in general there is a lack of logging in the core activities (at least in Elsa 1, haven't yet checked Elsa 2). 

Maybe give the users an option to define pre-and post-logs for the activities? That way they can define which variable they want to track before and after the activity got executed.

**Comment by craigfowler** [comment]
if it doesn't already exist, logging unhandled exceptions & stack traces (coming from activity execution) would be really important here.  I didn't check whether it's already in Elsa 2 though.

**Comment by sfmskywalker** [comment]
Elsa 2 has attained this feature fairly recently:

- Error information + stack trace are stored in the Workflow Instance.
- The faulting activity is logged in the Workflow Execution Log store as well (with a message of `"Faulted"`).

Elsa 2 also logs `"Executing"` and `"Executed"` lifecycle events of a given activity (including the source file + line number when the workflow blueprint originates from a workflow definition coded with the workflow builder API).

And I think it would be even more helpful if these events also record activity input & output values. The only concern I have right now is, what to do if the objects are big - depending on the workflow definition, the execution log could grow large really quick.

Perhaps we could indeed give the user a control to opt-out of automatic input/output logging as oer @mohdali 's suggestion. Perhaps by blacklisting certain types that are known to be big when serialized to JSON. Or perhaps a configurable setting in Elsa that controls the maximum number of bytes input & output is allowed to be stored. A combination might work well too.

Additionally, individual activity implementations such as `SetVariable` can of course add entries to the execution log themselves too. For instance, when `SetVariable` executes, it could add a new entry named e.g. `"Variable"`, with its payload being the serialized value.

It looks like there are a number of ways to skin this cat, so I suggest to start with a simple approach of just adding **input** to the **Executing** events and **output** to the *Executed* events. Having this alone will already improve the ability to troubleshoot workflows.

**Comment by craigfowler** [comment]
Yeah, another consideration when logging is _accidental logging of sensitive/confidential data_, particularly Personally Identifiable Information (PII).  I know for example that [Identity Server](https://identityserver.io/) intentionally has a config toggle which stops it from logging PII.  It's all good when you're logging system events etc, but once you start logging things like variables, you can easily catch things which shouldn't really be logged and it's down to the implementer as what goes in their variables.

The risk is that the owner of the app using Elsa might put together some great security for their persistence backend, protecting their sensitive data.  If copies of the data are being saved to a log that they didn't think to secure as tightly, then oops, security hole.

**Comment by sfmskywalker** [comment]
That's a very good and important point. Perhaps this variable/input/output logging needs to be an opt-in setting. When enabled, everything will be logged, but go through a filter allowing the developer to mask PII.

**Comment by mohdali** [comment]
@sfmskywalker what we used to have in an older software is two properties for each activities:
1- LogBeforeStart
2- LogAfterComplete

Each of these properties can be assigned an expression. So the user can fully customize the log message and print the variables they want, or the timestamp, or anything basically.

I think that approach could be useful for Elsa also. And maybe with the new approaches of the structured logging it can be further enhanced.

**Comment by sfmskywalker** [comment]
Interesting. We might add this to the same category of properties common to all activities, like whether the activity should cause the workflow to be persisted before / after executing, whether to load the workflow context and whether it should persist its output. In addition, it could have these 2 properties you mentioned.

**Comment by sfmskywalker** [comment]
I’m gonna propose we move this issue out of the milestone. Although it’s useful to have, it’s not really crucial for V2 to function.

**Comment by craigfowler** [comment]
Glad you said that! I was going to make the same suggestion! Part of this is down to a technical issue:

> How should we serialise an `IActivityExecutionResult`?

Granted, most of the time it will be an instance of a well-known result class, but if the consuming app developer has written their own, it could be almost anything. It's very hard to make assumptions about how we can meaningfully represent it in a log.

**Comment by sfmskywalker** [comment]
Great! Then let’s move this to the backlog.

Regarding your question, I don’t know. My initial thought wasn’t even to serialize the activity execution result, just an activity’s output and outcome. But your question made me realize that we might want to store things about the entire action result, not just OutputResult and OutcomeResult. In other words, we need to think this through some more.

**Comment by SeriousHulk** [comment]
What About Any Custom messages user want to Log along with any Activity? How can we add custom log messages in Elsa 2?

**Comment by sfmskywalker** [comment]
You should be able to log custom messages in Elsa 2 using the activity execution context, but the designer is currently hardcoded to filter by Started and Completed events - it uses these to calculate the timestamp deltas.

Elsa 3 has a better way of dealing with this by simply displaying all events, of any kind, including custom ones.
